### PR TITLE
feat: load jina auth token from environment

### PR DESCRIPTION
**Pull request by gmastrapas**
Goals:

- Enable passing Jina authentication token from the environment for `da.push` and `DocumentArray.pull`. Uses key `JINA_AUTH_TOKEN`
- [ ] check and update documentation, if required. See [guide](https://github.com/jina-ai/jina/blob/master/CONTRIBUTING.md#documentation-guidelines)

### CODE REVIEW

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/unit/array/mixins/test_pushpull.py [hunk 1]
```diff
@@ -179,7 +181,40 @@ def test_api_authorization_header(mocker, monkeypatch, tmpdir):
...
+
+
+def test_api_authorization_header_from_env(mocker, monkeypatch):
+    from docarray.array.mixins.io.pushpull import _get_hub_config, _get_auth_token
+
+    _get_hub_config.cache_clear()
+    _get_auth_token.cache_clear()
+
+    token = 'test-auth-token'
+    os.environ['JINA_AUTH_TOKEN'] = token
```
[the remark this answers is not in the document]
Comment (gmastrapas) [reply tests/unit/array/mixins/test_pushpull.py:204]: I unset the env var in the same tests

[the remark this answers is not in the document]
Comment (JoanFM) [reply tests/unit/array/mixins/test_pushpull.py:204]: yes, but what if there is an Exception in the middle, then this part of the code is not executed

[the remark this answers is not in the document]
Comment (gmastrapas) [reply tests/unit/array/mixins/test_pushpull.py]: ok makes sense

File: tests/unit/array/mixins/test_pushpull.py [hunk 2]
```diff
@@ -188,3 +192,32 @@ def test_api_authorization_header(mocker, monkeypatch, tmpdir):
...
+    _get_hub_config.cache_clear()
+    _get_auth_token.cache_clear()
+
+    assert mock.call_count == 3  # 1 for push, 1 for pull, 1 for download
+
+    _, push_kwargs = mock.call_args_list[0]
+    _, pull_kwargs = mock.call_args_list[1]
+
+    assert push_kwargs['headers'].get('Authorization') == f'token test-auth-token'
+    assert pull_kwargs['headers'].get('Authorization') == f'token test-auth-token'
```
Comment (alaeddine-13) [inline tests/unit/array/mixins/test_pushpull.py:223]:
```suggestion
    assert push_kwargs['headers'].get('Authorization') == 'token test-auth-token'
    assert pull_kwargs['headers'].get('Authorization') == 'token test-auth-token'
```

File: docarray/array/mixins/io/pushpull.py [hunk 3]
```diff
@@ -27,6 +27,14 @@ def _get_hub_config() -> Optional[Dict]:
             return json.load(f)
 
 
+@lru_cache()
+def _get_auth_token() -> Optional[str]:
+    _hub_config = _get_hub_config()
+    _config_auth_token = _hub_config.get('auth_token') if _hub_config else None
+    _env_auth_token = os.environ.get('JINA_AUTH_TOKEN')
+    return _env_auth_token or _config_auth_token
```
Comment (alaeddine-13) [inline docarray/array/mixins/io/pushpull.py:35]:
```suggestion
    hub_config = _get_hub_config()
    config_auth_token = hub_config.get('auth_token') if hub_config else None
    env_auth_token = os.environ.get('JINA_AUTH_TOKEN')
    return env_auth_token or config_auth_token
```

File: tests/unit/array/mixins/test_pushpull.py [hunk 4]
```diff
@@ -188,3 +192,32 @@ def test_api_authorization_header(mocker, monkeypatch, tmpdir):
 
     assert push_kwargs['headers'].get('Authorization') == f'token {token}'
     assert pull_kwargs['headers'].get('Authorization') == f'token {token}'
+
+
+@pytest.mark.parametrize(
+    'set_env_vars', [{'JINA_AUTH_TOKEN': 'test-auth-token'}], indirect=True
+)
+def test_api_authorization_header_from_env(mocker, monkeypatch, set_env_vars):
```
Comment (alaeddine-13) [inline tests/unit/array/mixins/test_pushpull.py:200]: let's just add one last test that asserts the priority in case both are set

**Review by JoanFM — APPROVED** [review]

**Review by alaeddine-13 — APPROVED** [review]

**Pull request merged**